### ISSUE TITLE
'?' appearing in scientific names extracted using dwc-stream

### ISSUE THREAD

**Issue by pgasu** [bio-guoda/preston#167]
I extracted GBIF archive (ver 38) data using dwc-stream, and the scientific names seem to have '?' character in a lot of occurrences. If i try to quantify almost 1% (47K) of unique mammal names (4.7M) have this special character. I am not sure if it is a common issue with the source data, or is it something happening during Preston archiving or extracting information. May be some guidance how we can approach this issue would be helpful.

**Comment by jhpoelen** [comment]
@pgasu thanks for sharing!

I am eager to get to the source of the "?" mystery. 

Do you happen to have some example along with the content ids of the archives they came from? A json snippet generated by ```preston dwc-stream``` should contain that information.

**Comment by jhpoelen** [comment]
Thanks for sharing the examples! 

one example suggestions that the ```?``` are hard coded in the provided darwin core archive:

```
$ preston cat --remote https://deeplinker.bio 'line:zip:hash://sha256/b40f806d6fd31a6c3edc81c29db87f85c4b2beafc2da3182fb8902e2152f2091!/taxon.txt!/L101'
[https://deeplinker.bio/b...2da3182fb8902e2152f2091] 100.0% of 17 kB at 0.67 MB/s completed in < 1 minute
4105847	4105847	Glyceria nemoralis Uechtr. & K?rn.	Plantae	SPECIES
```

If you'd like, you can contact the curators of the dataset using the contact info inside the dwca:

```
$ preston cat 'zip:hash://sha256/b40f806d6fd31a6c3edc81c29db87f85c4b2beafc2da3182fb8902e2152f2091!/eml.xml'
<eml:eml xmlns:eml="eml://ecoinformatics.org/eml-2.1.1"
         xmlns:dc="http://purl.org/dc/terms/"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="eml://ecoinformatics.org/eml-2.1.1 http://rs.gbif.org/schema/eml-gbif-profile/1.1/eml.xsd"
         packageId="http://gbif.ru:8080/ipt/resource?id=ptz_flora/v1.0" system="http://gbif.org" scope="system"
         xml:lang="eng">

<dataset>
  <alternateIdentifier>http://gbif.ru:8080/ipt/resource?r=ptz_flora</alternateIdentifier>
  <title xml:lang="eng">Vascular plants checklist of Prioksko-Terrasny Nature Reserve, European Russia</title>
      <creator>
    <individualName>
        <givenName>Yuri</givenName>
      <surName>Alekseev</surName>
    </individualName>
    <organizationName>Lomonosov Moscow State University</organizationName>
    <positionName>Associate Professor</positionName>
    <address>
        <city>Moscow</city>
        <country>RU</country>
    </address>
    <electronicMailAddress>alinaaksenova@yandex.ru</electronicMailAddress>
      </creator>
      <creator>
    <individualName>
        <givenName>Tatiana</givenName>
      <surName>Sycheva</surName>
    </individualName>
    <organizationName>Prioksko-Terrasny Nature Reserve</organizationName>
    <positionName>Researcher</positionName>
    <address>
        <city>Danki</city>
        <country>RU</country>
    </address>
    <electronicMailAddress>tanya.sycheva.heyhey@gmail.com</electronicMailAddress>
      </creator>
      <creator>
    <individualName>
        <givenName>Larisa</givenName>
      <surName>Denisova</surName>
    </individualName>
    <organizationName>Prioksko-Terrasny Nature Reserve</organizationName>
    <positionName>Researcher</positionName>
    <address>
        <city>Danki</city>
        <country>RU</country>
    </address>
      </creator>
      <metadataProvider>
    <individualName>
        <givenName>Tatiana</givenName>
      <surName>Sycheva</surName>
    </individualName>
    <organizationName>Prioksko-Terrasny Nature Reserve</organizationName>
    <positionName>Researcher</positionName>
    <address>
        <city>Danki</city>
        <country>RU</country>
    </address>
    <electronicMailAddress>tanya.sycheva.heyhey@gmail.com</electronicMailAddress>
      </metadataProvider>
      <associatedParty>
    <individualName>
        <givenName>Natalya</givenName>
      <surName>Ivanova</surName>
    </individualName>
    <electronicMailAddress>natalya.dryomys@gmail.com</electronicMailAddress>
    <role>user</role>
      </associatedParty>
  <pubDate>
      2019-08-17
  </pubDate>
  <language>eng</language>
  <abstract>
    <para>Present checklist contains data on vascular plants of Prioksko-Terrasny Nature Reserved, located in European Russia (Russian Federation, south of Moscow region). The first annotated list of the reserve’s flora was released in 1958 by P. Smirnov, which also includes 30-year observations of flora. At the moment, 913 vascular plants species detected in the Reserve and it is updated constantly.</para>
  </abstract>
      <keywordSet>
            <keyword>Checklist</keyword>
        <keywordThesaurus>GBIF Dataset Type Vocabulary: http://rs.gbif.org/vocabulary/gbif/dataset_type.xml</keywordThesaurus>
      </keywordSet>
      <keywordSet>
            <keyword>Inventorythematic</keyword>
        <keywordThesaurus>GBIF Dataset Subtype Vocabulary: http://rs.gbif.org/vocabulary/gbif/dataset_subtype.xml</keywordThesaurus>
      </keywordSet>
  <intellectualRights>
    <para>This work is licensed under a <ulink url="http://creativecommons.org/licenses/by/4.0/legalcode"><citetitle>Creative Commons Attribution (CC-BY) 4.0 License</citetitle></ulink>.</para>
  </intellectualRights>
  <maintenance>
    <description>
      <para></para>
    </description>
    <maintenanceUpdateFrequency>unkown</maintenanceUpdateFrequency>
  </maintenance>

      <contact>
    <individualName>
        <givenName>Tatiana</givenName>
      <surName>Sycheva</surName>
    </individualName>
    <organizationName>Prioksko-Terrasny Nature Reserve</organizationName>
    <positionName>Researcher</positionName>
    <address>
        <city>Danki</city>
        <country>RU</country>
    </address>
    <electronicMailAddress>tanya.sycheva.heyhey@gmail.com</electronicMailAddress>
      </contact>
  <project id="Russia2019_03">
    <title>Chronicles of Nature of Russian protected areas: Digitization and data mobilization</title>
      <personnel>
        <individualName>
            <givenName>Yuri</givenName>
          <surName>Buyvolov</surName>
        </individualName>
        <role>curator</role>
      </personnel>
  </project>
</dataset>
  <additionalMetadata>
    <metadata>
      <gbif>
          <dateStamp>2019-08-17T01:23:49.410+03:00</dateStamp>
          <hierarchyLevel>dataset</hierarchyLevel>
      </gbif>
    </metadata>
  </additionalMetadata>
</eml:eml>
```

**Comment by jhpoelen** [comment]
Here's some more examples from that same archive:

```
$ preston cat 'zip:hash://sha256/b40f806d6fd31a6c3edc81c29db87f85c4b2beafc2da3182fb8902e2152f2091!/taxon.txt' | grep "\?"
4105847	4105847	Glyceria nemoralis Uechtr. & K?rn.	Plantae	SPECIES
8042333	8042333	Dactylorhiza cruenta (O.F.M?ll.) So?	Plantae	SPECIES
2810677	2810677	Dactylorhiza fuchsii (Druce) So?	Plantae	SPECIES
8248593	8248593	Dactylorhiza incarnata (L.) So?	Plantae	SPECIES
8179794	8179794	Dactylorhiza maculata (L.) So?	Plantae	SPECIES
2810532	2810532	Dactylorhiza traunsteineri (Saut. ex Rchb.) So?	Plantae	SPECIES
7911993	7911993	Rumex pseudonatronatus (Borb?s) Murb.	Plantae	SPECIES
7602511	7602511	Viola braunii Borb?s	Plantae	SPECIES
5663309	5663309	Viola ruprechtiana Borb?s	Plantae	SPECIES
```

**Comment by jhpoelen** [comment]
related dataset as indexed by GBIF 

https://www.gbif.org/dataset/954f6aaa-c2e9-41e4-81ca-04eb31912fef

interestingly, the index taxon record on GBIF https://www.gbif.org/species/159152139 notes as an issue "name partially parsed" and appears to have chopped off the name including the ```?``` and following (e.g., ````(Glyceria nemoralis Uechtr. & K?rn.) -[:translated by GBIF to]-> (Glyceria nemoralis Uechtr. & K)```

![Screenshot from 2022-04-20 16-31-26](https://user-images.githubusercontent.com/1084872/164326430-eacb5285-5b29-4ba6-a4b8-314363c75901.png)
![Screenshot from 2022-04-20 16-30-39](https://user-images.githubusercontent.com/1084872/164326431-8019eaf0-ac06-4a3e-a9ca-c5b90ff9db1b.png)
![Screenshot from 2022-04-20 16-32-41](https://user-images.githubusercontent.com/1084872/164326918-9085d43f-1eea-46d7-9fe8-e2e4f8604758.png)

**Comment by jhpoelen** [comment]
Neat to see how we can trace the provenance of name with ```?``` to their sources with surgical precision. And, learn how the names propagate into other infrastructures like GBIF. 

@n8upham @pgasu curious to hear your ideas on your preferred way to deal with these suspicious names.

**Comment by n8upham** [comment]
Hey @jhpoelen and @pgasu -- It looks like there is a multi-part solution:

- Some of these ? are from UTF character conversion problems, e.g., "Cheilosia grossa (Fall?n, 1817)" should refer to "Fallén" see https://en.wikipedia.org/wiki/Cheilosia_grossa -- seems to mostly affect authority names

- Others, like this hybrid "Euphrasia stricta x Euphrasia micrantha ???" are due to actual ambiguity in the data record, meaning that the tag of the specimen likely had ??? and that was transcribed to the DB

Are there other examples that don't fit either of those?

**Comment by pgasu** [comment]
@jhpoelen and @n8upham - My main concern was to figure out whether the error lies in the source data or our data pipeline.  To me it looks like a character encoding issue emerged at some stage while reading csv/txt files. To give you some numbers - such errors exist in ~1000 datasets, and number of occurrences affected are ~87k for 47k unique mammal names.

**Comment by jhpoelen** [comment]
@pgasu thanks for giving a sense of the scale of the issue. How many occurrences did you consider? 

Also, I have not yet seen any evidence for that the "?" are introduced on reading from the dwca. Instead, the examples I've seen suggest that the "?"'s were introduce on producing and publishing the dwca. In other words, unless I see other examples, I believe that most "?"'s are hard-coded in the archives. Please share some examples if you have them, because I can totally wrong about this.

**Comment by pgasu** [comment]
@jhpoelen I agree that '?'s are introduced during production and publishing of dwca on GBIF -  that's what I meant in my previous comment.
Actually, I just found out that it is a known problem https://discourse.gbif.org/t/100-gbif-datasets-improved/3042 - check `Data item formatting problems` and `Character problems in data items` - may be these are already fixed in the later versions.

**Comment by pgasu** [comment]
> How many occurrences did you consider?
There were almost 227 million total occurrences and 4.7 million unique mammal names. These numbers include a lot of duplication because of formatting, inconsistent use of author name (with or without parentheses), and other human errors.

**Comment by jhpoelen** [comment]
> Actually, I just found out that it is a known problem https://discourse.gbif.org/t/100-gbif-datasets-improved/3042 - check Data item formatting problems and Character problems in data items - may be these are already fixed in the later versions.

Great find! We should be able to check whether the issue was resolved later, by re-running the analysis on newer snapshots and counting "?"'s .

**Comment by jhpoelen** [comment]
@pgasu in reviewing open issues, I found this issue related to the "?" in scientific name you found earlier this year. According to the comments, the "?" come from the datasets themselves, and are not introduced by Preston or Preston's dwc-stream. 

Closing issue. Happy to re-open issue if you have remaining thoughts on the subject.

Thanks again for raising this issue and taking the time to document it.
